prow/plugins/verify-owners/verify-owners.go

```diff
@@ -0,0 +1,171 @@
+	"k8s.io/test-infra/prow/repoowners"
+)
+
+const (
+	pluginName     = "verify-owners"
+	ownersFileName = "OWNERS"
+)
+
+var (
+	invalidOwnersLabel = "do-not-merge/invalid-OWNERS"
+)
+
+func init() {
+	plugins.RegisterPullRequestHandler(pluginName, handlePullRequest, helpProvider)
+}
+
+func helpProvider(config *plugins.Configuration, enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	return &pluginhelp.PluginHelp{
+			Description: "The owners plugin validates OWNERS files if they are modified in a PR. On failure it automatically adds a 'do-not-merge/invalid-OWNERS' label to the PR, and a review comment on the incriminated file.",
+		},
+		nil
+}
+
+type ownersClient interface {
+	LoadRepoOwners(org, repo, base string) (repoowners.RepoOwnerInterface, error)
+}
+
+type githubClient interface {
+	AddLabel(org, repo string, number int, label string) error
+	CreateReview(org, repo string, number int, r github.DraftReview) error
+	GetPullRequestChanges(org, repo string, number int) ([]github.PullRequestChange, error)
+}
+
+func handlePullRequest(pc plugins.PluginClient, pre github.PullRequestEvent) error {
+	if pre.Action != github.PullRequestActionOpened && pre.Action != github.PullRequestActionReopened && pre.Action != github.PullRequestActionSynchronize {
+		return nil
+	}
+	return handle(pc.GitHubClient, pc.GitClient, pc.Logger, &pre, pc.PluginConfig.Owners.LabelsBlackList)
+}
+
+func handle(ghc githubClient, gc *git.Client, log *logrus.Entry, pre *github.PullRequestEvent, labelsBlackList []string) error {
+	org := pre.Repo.Owner.Login
+	repo := pre.Repo.Name
+	wrongOwnersFiles := map[string]error{}
+
+	// Get changes.
+	changes, err := ghc.GetPullRequestChanges(org, repo, pre.Number)
+	if err != nil {
+		return fmt.Errorf("error getting PR changes: %v", err)
+	}
+
+	// List modified OWNERS files.
+	var modifiedOwnersFiles []string
+	for _, change := range changes {
+		if filepath.Base(change.Filename) == ownersFileName {
+			modifiedOwnersFiles = append(modifiedOwnersFiles, change.Filename)
+		}
+	}
+	if len(modifiedOwnersFiles) == 0 {
+		return nil
+	}
+
+	// Clone the repo, checkout the PR.
+	r, err := gc.Clone(pre.Repo.FullName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := r.Clean(); err != nil {
+			log.WithError(err).Error("Error cleaning up repo.")
+		}
+	}()
+	if err := r.CheckoutPullRequest(pre.Number); err != nil {
+		return err
+	}
+
+	// Check each OWNERS file.
+	for _, f := range modifiedOwnersFiles {
+		// Try to load OWNERS file.
+		path := filepath.Join(r.Dir, f)
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to read %s.", path)
+			return nil
+		}
+		var approvers []string
+		var labels []string
+		simple, err := repoowners.ParseSimpleConfig(b)
+		if err != nil || simple.Empty() {
+			full, err := repoowners.ParseFullConfig(b)
+			if err != nil {
+				wrongOwnersFiles[f] = fmt.Errorf("error occurred parsing the OWNERS file in %s: %v", f, err)
+				break
```

Comment:
Why `break` after adding to `wrongOwnersFiles`? Don't we want to give as much feedback as we can in one go?

Reply:
I thought it would not make sense to try more tests if that OWNERS file is already invalid. Especially if we cannot parse it, how to validate the labels and approvers?

Reply:
I mean more that we should not break out of the loop over every file -- `continue` on to the next file, perhaps, but we should attempt to validate each file that was changed.

Reply:
You're right... I wanted to `continue` here, otherwise it makes no sense to check the length of wrongOwnersFiles. Thanks for spotting it!